### DISCUSSION TITLE
Fate of feature/pvh branch

### DISCUSSION THREAD

**Discussion by nichtverstehen** [firecracker-microvm/firecracker#4611] [General]
Hi folks!

I want to inquire on the situation with the PVH boot protocol support in Firecracker.

Eight months ago @cperciva implemented and @roypat reviewed changes to make Firecracker support PVH boot protocol correctly. The changes were accepted, but landed in a feature branch that didn't see updates since then:
https://github.com/firecracker-microvm/firecracker/tree/feature/pvh

I stumbled upon them while experimenting with running SerenityOS on Firecracker. It would be fun to make it work, but as it's a non-Linux OS, it is too tricky to implement Linuxboot protocol.

I rebased the branch and with minor changes PVH boot works! The rebased changes are here:
https://github.com/nichtverstehen/firecracker/tree/293417a2f35190749689926daf7c8bf7486f631a

What would it take to merge `feature/pvh` into mainline?

**Comment by roypat** [comment]
Hi @nichtverstehen,
Thanks for reaching out about this! It's always cool to see people running different things on Firecracker :D

There were some remaining issues with PVH-boot support that we track in https://github.com/firecracker-microvm/firecracker/pull/4073. The functional blocker was that initrds were not working with PVH, and neither us nor Colin had time to look into that since. However, I saw your PR on linux-loader (https://github.com/rust-vmm/linux-loader/pull/179), which fixes exactly that! So with that resolved, we'll still need to do some boottime benchmarks, and allocate some time to on our side to get the feature across the finish line (internal security review, etc.). So there isn't much, but we also don't currently know when we'll be able to get back to it, especially because we do not have a strong usecase for it currently :(
Best, 
Patrick

**Comment by cperciva** [comment]
Awesome!  Sorry about abandoning this work -- I took over as FreeBSD release engineer and didn't have time to finish polishing this Firecracker.

@roypat In case it makes any difference to prioritization, I believe that NetBSD boots in Firecracker with PVH, too.  So I think that's three platforms which will start working once PVH boot support is merged. ;-)

**Comment by nichtverstehen** [comment]
Great to hear and happy to see it moving! Let me know if I can help with something. If you don't mind I'll keep nagging here once in a while :).

You're right that PVH boot is probably not quite as popular for production use these days. It's a bit of a funky artefact. What would be actually be more useful is supporting Multiboot (https://en.wikipedia.org/wiki/Multiboot_specification). That would enable some more practical cases, like booting L4 or Zircon/Fuchsia. (I suppose BSDs could boot on multiboot too).

But we have what we have :).
